base/threadingconstructs.jl

```diff
@@ -175,9 +175,57 @@ end
 function _threadsfor(iter, lbody, schedule)
     lidx = iter.args[1]         # index
     range = iter.args[2]
+    esc_range = esc(range)
+    func = if schedule === :greedy
+        greedy_func(esc_range, lidx, lbody)
+    else
+        default_func(esc_range, lidx, lbody)
+    end
     quote
         local threadsfor_fun
-        let range = $(esc(range))
+        $func
+        if $(schedule === :greedy || schedule === :dynamic || schedule === :default)
+            threading_run(threadsfor_fun, false)
+        elseif ccall(:jl_in_threaded_region, Cint, ()) != 0 # :static
+            error("`@threads :static` cannot be used concurrently or nested")
+        else # :static
+            threading_run(threadsfor_fun, true)
+        end
+        nothing
+    end
+end
+
+function greedy_func(itr, lidx, lbody)
+    quote
+        let r = Ref{Task}(), c = Channel{eltype($itr)}(threadpoolsize(),taskref=r) do ch
+            for item in $itr
+                put!(ch, item)
+            end
+        end
+        function threadsfor_fun(tid)
+            t = r[]
+            while !istaskstarted(t)
+                yield(t)
+            end
+            while isready(c)
+                item = try
+                    take!(c)
+                catch e
+                    # should technically never happen, but could?!
+                    # we don't have a non-blocking `take!`
+                    rethrow(e)
+                end
```

Comment:
This part in particular I'm not sure about. The thinking is that the channel could become empty between `isready` and `take!`, resulting in an error. One way to "safeguard" that is to check `e isa InvalidStateException` and then `break`, as the iteration is successfully finished. Thoughts, anyone?

Reply:
This has happened in CI, and is now done by no longer having this broken code, relying on iteration over channels instead (which does the right thing already).